sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/caches/AsyncCacheNonBlocking.java

```diff
@@ -0,0 +1,228 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import reactor.core.publisher.Mono;
+
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.Function;
+
+public class AsyncCacheNonBlocking<TKey, TValue> {
+    private final Logger logger = LoggerFactory.getLogger(AsyncCacheNonBlocking.class);
+    private final ConcurrentHashMap<TKey, AsyncLazyWithRefresh<TValue>> values;
+    private final IEqualityComparer<TKey> keyEqualityComparer;
+
+    public AsyncCacheNonBlocking(ConcurrentHashMap<TKey, AsyncLazyWithRefresh<TValue>> values, IEqualityComparer<TKey> keyEqualityComparer) {
+        this.values = values;
+        this.keyEqualityComparer = keyEqualityComparer;
+    }
+
+    public AsyncCacheNonBlocking(IEqualityComparer<TKey> keyEqualityComparer) {
+        this(new ConcurrentHashMap<>(), keyEqualityComparer);
+    }
+
+    public AsyncCacheNonBlocking() {
+        this((value1, value2) -> {
+            if (value1 == value2)
+                return true;
+            if (value1 == null || value2 == null)
+                return false;
+            return value1.equals(value2);
+        });
+    }
+
+    /**
+     * <summary>
+     * <para>
+     * Gets value corresponding to <paramref name="key"/>.
+     * </para>
+     * <para>
+     * If another initialization function is already running, new initialization function will not be started.
+     * The result will be result of currently running initialization function.
+     * </para>
+     * <para>
+     * If previous initialization function is successfully completed it will return the value. It is possible this
+     * value is stale and will only be updated after the force refresh task is complete.
+     * </para>
+     * <para>
+     * Force refresh is true:
+     * If the key does not exist: It will create and await the new task
+     * If the key exists and the current task is still running: It will return the existing task
+     * If the key exists and the current task is already done: It will start a new task to get the updated values.
+     * Once the refresh task is complete it will be returned to caller.
+     * If it is a success the value in the cache will be updated. If the refresh task throws an exception the key will be removed from the cache.
+     * </para>
+     * <para>
+     * If previous initialization function failed - new one will be launched.
+     * </para>
+     * </summary>
+     */
+    public Mono<TValue> getAsync(
+        TKey key,
+        Function<TValue, Mono<TValue>> singleValueInitFunc,
+        boolean forceRefresh) {
+
+        AsyncLazyWithRefresh<TValue> initialLazyValue = values.get(key);
+        if (initialLazyValue != null) {
+            logger.debug("cache[{}] exists", key);
+            return initialLazyValue.getValueAsync().flatMap(value -> {
```

Comment:
How are we going to distinguish between the task is still running vs the task has finished. This is a question for this related comments:
```
 * If the key exists and the current task is still running: It will return the existing task
 * If the key exists and the current task is already done: It will start a new task to get the updated values.
```